SingularityService/src/main/java/com/hubspot/singularity/data/LoggingCuratorFramework.java

```diff
@@ -0,0 +1,312 @@
+import org.apache.curator.framework.api.GetConfigBuilder;
+import org.apache.curator.framework.api.GetDataBuilder;
+import org.apache.curator.framework.api.ReconfigBuilder;
+import org.apache.curator.framework.api.RemoveWatchesBuilder;
+import org.apache.curator.framework.api.SetACLBuilder;
+import org.apache.curator.framework.api.SetDataBuilder;
+import org.apache.curator.framework.api.SyncBuilder;
+import org.apache.curator.framework.api.UnhandledErrorListener;
+import org.apache.curator.framework.api.transaction.CuratorMultiTransaction;
+import org.apache.curator.framework.api.transaction.CuratorTransaction;
+import org.apache.curator.framework.api.transaction.TransactionOp;
+import org.apache.curator.framework.imps.CuratorFrameworkState;
+import org.apache.curator.framework.listen.Listenable;
+import org.apache.curator.framework.schema.SchemaSet;
+import org.apache.curator.framework.state.ConnectionStateErrorPolicy;
+import org.apache.curator.framework.state.ConnectionStateListener;
+import org.apache.curator.utils.EnsurePath;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.server.quorum.flexible.QuorumVerifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoggingCuratorFramework implements CuratorFramework {
+  private final CuratorFramework curator;
+
+  private final ConcurrentMap<String, Integer> counters;
+
+  private final Long interval = 60000L * 3; // 3 minutes
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+    LoggingCuratorFramework.class
+  );
+
+  private class ClearCounter extends TimerTask {
+
+    @Override
+    public void run() {
+      logAndClear();
+    }
+  }
+
+  @Inject
+  public LoggingCuratorFramework(CuratorFramework curator) {
+    this.curator = curator;
+    counters = new ConcurrentHashMap<>();
+    Timer timer = new Timer();
+
+    timer.schedule(new ClearCounter(), interval, interval);
+  }
+
+  private String getCaller() {
+    StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
+    long threadId = Thread.currentThread().getId();
+
+    int levelInStack = 0;
+    String className = "";
+    for (int i = 0; i < stackTraceElements.length; i++) {
+      String longClassName = stackTraceElements[i].getClassName();
+      className = longClassName.substring(longClassName.lastIndexOf(".") + 1);
+
+      if (
+        !(
+          className.equals("Thread") ||
+          className.equals("LoggingCuratorFramework") ||
+          className.equals("CuratorManager") ||
+          className.equals("CuratorAsyncManager")
+        )
+      ) {
+        levelInStack = i;
+        break;
+      }
+    }
+
+    return (
+      threadId +
+      ": " +
+      className +
+      "\'s " +
+      stackTraceElements[levelInStack].getMethodName()
+    );
+  }
+
+  private void setCounter() {
+    String caller = getCaller();
+    int counter = counters.getOrDefault(caller, 0);
+    counters.put(caller, ++counter);
```

Comment:
getting and setting in separate operations won't be atomic, something like `counters.compute(key, (k, v) -> v == null ? 1 : v+1)` should do the trick.

Reply:
nit — may want to change `setCounter` with `incCounter` or similar for clarity